### ISSUE TITLE
Generation context does not restore state after errors

### ISSUE THREAD

**Issue by RerankerGuo** [huggingface/trl#6659]
### Reproduction

`_unwrap_model_for_generation` temporarily disables gradient checkpointing and, for gathered DeepSpeed ZeRO-3 generation, removes optimizer hooks. The cleanup statements are placed after `yield` without `finally` blocks. If generation raises inside the context, Python rethrows the exception at `yield` and skips both cleanup paths.

The gradient-checkpointing leak can be reproduced without distributed hardware:

```python
import types
from unittest.mock import patch

from trl.models.utils import _unwrap_model_for_generation


model = types.SimpleNamespace(is_gradient_checkpointing=True)
model.gradient_checkpointing_disable = lambda: setattr(model, "is_gradient_checkpointing", False)
model.gradient_checkpointing_enable = lambda: setattr(model, "is_gradient_checkpointing", True)
accelerator = types.SimpleNamespace(unwrap_model=lambda value: value)

try:
    with patch("trl.distributed.DistributedBackend", return_value=types.SimpleNamespace(is_zero3=False)):
        with _unwrap_model_for_generation(model, accelerator):
            raise RuntimeError("generation failed")
except RuntimeError:
    pass

print(model.is_gradient_checkpointing)
```

Output on `main` at `38b4b810`:

```text
False
```

The same control flow skips `add_hooks(model)` when an exception is raised from the gathered ZeRO-3 branch after `remove_hooks(model)`.

Temporary generation state should be restored whether the body exits normally or raises. The original generation exception should continue to propagate.

### System Info

- Platform: macOS-26.5.1-arm64-arm-64bit
- Python version: 3.12.13
- TRL base commit: 38b4b810
- PyTorch version: 2.13.0
- accelerator(s): MPS
- Transformers version: 5.14.1
- Accelerate version: 1.14.0
- Datasets version: 5.0.1
- HF Hub version: 1.26.0
- DeepSpeed version: not installed (the ZeRO-3 lifecycle is reproduced with a lightweight module stub)

### Checklist

- [x] I have checked that my issue isn't already filed.
- [x] I have included my system information.
- [x] Any code provided is minimal, complete, and reproducible.
- [x] Any code provided is properly formatted in code blocks.
- [x] Any traceback provided is complete.

**Comment by YeonwooSung** [comment]
I'll take this. Wrapping the yield in `_unwrap_model_for_generation` with `try`/`finally` so gradient checkpointing and ZeRO-3 hooks are restored if generation raises.
